docs/topics/tasks.txt

```diff
@@ -0,0 +1,337 @@
+  validation can be disabled by setting this to an empty list.
+
+Asynchronous support
+--------------------
+
+Django has developing support for asynchronous task backends.
+
+``django.tasks.backends.base.BaseTaskBackend`` has async variants of all base
+methods. By convention, the asynchronous versions of all methods are prefixed
+with ``a``. The arguments for both variants are the same.
+
+Retrieving backends
+-------------------
+
+Backends can be retrieved using the ``tasks`` connection handler::
+
+    from django.tasks import tasks
+
+    tasks["default"]  # The default backend
+    tasks["reserve"]  # Another backend
+
+The "default" backend is available as ``default_task_backend``::
+
+    from django.tasks import default_task_backend
+
+Backend features
+----------------
+
+Django's tasks framework is designed to be backend-agnostic. This means some
+backends may not implement the same features as others, such as executing ``async``
+functions.
+
+Depending on the backend configured, it may be necessary to gracefully degrade
+functionality, or use a system check to prevent the application from starting at all.
+
+To facilitate this, certain features can be checked on a backend:
+
+* :attr:`supports_defer <django.tasks.backends.base.BaseTaskBackend.supports_defer>`:
+  Can tasks be executed after a specific time using ``run_after``?
+
+* :attr:`supports_async_task <django.tasks.backends.base.BaseTaskBackend.supports_async_task>`:
+  Can ``async`` functions (coroutines) be used as task functions?
+
+* :attr:`supports_get_result <django.tasks.backends.base.BaseTaskBackend.supports_get_result>`:
+  Can a task's results be retrieved from another thread or process?
+
+Defining tasks
+==============
+
+Tasks are defined using the :meth:`django.tasks.task` decorator on a
+module-level function::
+
+    from django.tasks import task
+
+
+    @task()
+    def calculate_meaning_of_life(answer=42):
+        return answer
+
+Returned in a :class:`django.tasks.Task` instance.
+
+.. note::
+
+    All arguments and return values must be JSON serializable, and round-trip
+    as the same types.
+
+    This means complex types like model instances, as well as many built-in types
+    like ``datetime`` and ``tuple`` cannot be used in tasks without additional
+    conversion.
+
+The ``task`` decorator accepts a few keyword arguments to customize the task:
+
+* ``priority``: The priority of the task. Higher numbers will be executed sooner.
+
+* ``queue_name``: The name of the queue the task will be executed on
+
+* ``backend``: The name of the backend this task must use (as defined in
+  :setting:`TASKS`).
+
+* ``enqueue_on_commit``: Whether the task is enqueued when the current transaction
+  commits successfully, or enqueued immediately.
+
+These arguments correspond to attributes on the created
+:class:`Task <django.tasks.Task>`.
+
+By convention, tasks should be defined in a ``tasks.py`` file, however this is
+not enforced.
```

Comment:
(I think it _should_ be auto-imported, I think that's just a question of modifying `AppConfig`)

Reply:
Celery already uses the `tasks.py` convention, so there are going to be many projects that already have this file present that are unrelated to `django.tasks`.

I do foresee a need to get a list of all registered tasks, whether that happens by checking each installed app's `tasks.py` module for instances of `backend.task_class`, or adding a registration dictionary to the `@task` decorator. The use case would be something like scheduling a task from the admin portal, similar to how Celery Beat allows this by picking a registered task from a dropdown list.

Reply:
This is why I think it's valuable to have the Database backend here. If we don't auto-import `tasks.py`, but are recommending that we define tasks in `tasks.py`, and someone creates a task in `tasks.py` that gets queued in a view, but would be out of the import tree for a task worker, then the background worker won't know about the task to execute it! 

(That is to say, unless the database worker is in fact doing something like `import_string` on a task name... which I would find to be not ideal)

Anyways, big tent version of this:
- How does a background worker know about a task in `tasks.py`? Do we rely on the import tree hitting all tasks?
- Is the backend responsible for trawling import trees, rather than the `AppConfig`?

I am OK with any result now, but I think DX-wise the long term should have some form of auto-importing if we're recommending that tasks be defined there.

Reply:
I like the DX of auto-importing and it should be fine if we provide an escape hatch, should this behavior conflict with existing Celery code for some reason.

Celery provides `autodiscover_tasks()` where we can do something like `autodiscover_tasks(related_name='tasks2')` to have it auto-import the `tasks2.py` module or `tasks2` package of each Django app. We can provide something similar, or a `TASKS_` setting in `settings.py` to define the module/package path to use (which defaults to `tasks`).

Reply:
This can maybe be a feature for later down the road then. Don't want to increase the number of question marks if we can avoid it.

Reply:
When a task is saved to the store, its module path is saved with it, so `import_string` is used to find it. There's no need to know about all tasks in memory.